### PR TITLE
Add flag to disable dune diagnostics

### LINKED ISSUE #704: false negatives on unsaved files
Currently if you have an unsaved file that previously had an error and the error was removed but the file is still not saved ocamllsp show errors that don't exist anymore coming from dune.

A solution would be to disable dune errors on unsaved files.

### PR DESCRIPTION
**Pull request by EduardoRFS**
## Context

I like to work with incomplete files a lot and playing around with them before actually saving, going through many intermediary steps.

But due to how OCaml LSP behaves as soon as you save a file with a single error coming from dune, that error will be there until you save it again, which is quite annoying.

So I've been using for almost 2 years a patched version of the OCaml LSP without dune diagnostics, this is hopefully that but as an opt-in option.

## Patch

I tried to make the patch the least invasive, keeping all the features except the dune reporting inside of files, to do that I disable the **entire diagnostics loop**, this may be too much, but in the last years I never detected any issues.

## How to use

```json
{
  "ocaml.sandbox": {
    "kind": "global"
  },
  "ocaml.server.args": [
    "--no-dune-diagnostics"
  ]
}
```

## Related

- Closes #704

### CODE REVIEW

**Comment by EduardoRFS** [comment]
@tjdevries I thought about it, but I was not sure if OCaml LSP already had any options, but after checking it again, that seems to be the case, such as "extendedHover" and VSCode OCaml Platform even exposes that.

And it even already supports the on change, the only downside is that I need to implement that, but will do, also keeping the diagnostics_loop running is likely better.

**Comment by EduardoRFS** [comment]
This patch works, but it shines light in a likely bug, ocaml-lsp doesn't request the configs when starting the LSP, this means that it only detects the settings when changing them.

This seems to also explain why sometimes codelens is not loaded as the default is `codelens = Some { enable = false }`, I wonder if this is a `vscode-ocaml-platform` thing.

https://github.com/EduardoRFS/vscode-ocaml-platform/tree/no-dune

**Comment by voodoos** [comment]
Maybe another option would be to invalidate the diagnostics as soon as the buffer is edited ? (Not sure if that's allowed by the protocol however).

**Comment by EduardoRFS** [comment]
@voodoos I think this would be possible, but in general seems weird, if you're changing the file is likely that you're trying to solves some of those errors, removing the errors would make that harder for developers on a single screen.

**Comment by EduardoRFS** [comment]
I think the weird behavior of not pushing config was an issue with vscode-ocaml-platform and even if it is not, this patch seems to be safer overall.

https://github.com/ocamllabs/vscode-ocaml-platform/pull/1321

**Comment by voodoos** [comment]
> @voodoos I think this would be possible, but in general seems weird, if you're changing the file is likely that you're trying to solves some of those errors, removing the errors would make that harder for developers on a single screen.

Another way to think about it is that most compiler errors for the currently edited buffer are redundant with the ones already reported by Merlin. The most useful errors enabled by the dune-diagnostics are signature mismatches errors, that Merlin is not able to provide. 

We might want to only report these for the active buffer ? This would also solve the issue of having duplicated type errors when the compiler prints them differently by Merlin.

**Comment by EduardoRFS** [comment]
> The most useful errors enabled by the dune-diagnostics are signature mismatches errors, that Merlin is not able to provide.

I don't want to see those either, because again, if I make it match the signature without saving, it will be red for me. In fact I'm much rather have a false positive that I discover when trying to compile than a false negative that is annoying me while working.

But this is clearly a personal opinion and doesn't match other's experiences, so I would actually like the flag.

In my opinion, having dune diagnostics for any type error is just a failure of Merlin and we should work on that.

**Review by voodoos — COMMENTED** [review]
Made a first round of review with a few remarks. I believe there is some cleaning to do in the `Dune` module.

I am not opposed to adding a flag to let users choose if they want to use the feature or not. It's sad that the LSP protocol doesn't provide a standardize way to specify such simple options.

File: ocaml-lsp-server/src/config_data.ml [hunk 1]
```diff
@@ -145,11 +145,87 @@ module ExtendedHover = struct
...
+
 type t =
   { codelens : Lens.t Json.Nullable_option.t
         [@default None] [@yojson_drop_default ( = )]
   ; extended_hover : ExtendedHover.t Json.Nullable_option.t
         [@key "extendedHover"] [@default None] [@yojson_drop_default ( = )]
+  ; dune_diagnostics : DuneDiagnostics.t
+        [@key "duneDiagnostics"]
+        [@default DuneDiagnostics.{ enable = true }]
+        [@yojson_drop_default ( = )]
```
[the remark this answers is not in the document]
Comment (EduardoRFS) [reply ocaml-lsp-server/src/config_data.ml:228]: I don't know, my bet is that it was to be able to distinguish between default and null, but it doesn't matter in any of them, as they all treat both options the same, I can change the other ones also if you want

@aspeddro may know better

File: ocaml-lsp-server/src/dune.mli [hunk 2]
```diff
@@ -8,6 +8,9 @@ end
 
 type t
 
+(** enabled by default *)
+val enable_diagnostics : bool ref
```
[the remark this answers is not in the document]
Comment (EduardoRFS) [reply ocaml-lsp-server/src/dune.mli:12]: Yup, nice catch, my bad I forgot to review the new implementation.

File: ocaml-lsp-server/src/ocaml_lsp_server.ml [hunk 3]
```diff
@@ -702,7 +703,13 @@ let on_notification server (notification : Client_notification.t) :
     state
   | CancelRequest _ -> Fiber.return state
   | ChangeConfiguration req ->
-    let+ configuration = Configuration.update state.configuration req in
+    let* configuration = Configuration.update state.configuration req in
+    let+ () =
+      Diagnostics.set_report_dune_diagnostics
+        (State.diagnostics state)
+        ~report_dune_diagnostics:
+          state.configuration.data.dune_diagnostics.enable
```
[the remark this answers is not in the document]
Comment (EduardoRFS) [reply ocaml-lsp-server/src/ocaml_lsp_server.ml]: I would also like to not use mutation in general, but Diagnostics already had mutation, so I decided to go with that as the alternative would require quite a bit of work due to concurrency and already mutable state.

File: ocaml-lsp-server/docs/ocamllsp/config.md [hunk 4]
```diff
@@ -21,5 +21,12 @@ interface config {
   * @since 1.16
   */
   codelens: { enable : boolean }
+
+  /**
+  * Enable/Disable Dune diagnostics
+  * @default false
```
[the remark this answers is not in the document]
Comment (EduardoRFS) [reply ocaml-lsp-server/docs/ocamllsp/config.md]: Fixed

File: ocaml-lsp-server/src/config_data.ml (hunk 1 above)
[the remark this answers is not in the document]
Comment (EduardoRFS) [reply ocaml-lsp-server/src/config_data.ml:240]: To help here I moved all of them to the simpler format.

**Comment by EduardoRFS** [comment]
Tested this patch with the VSCode OCaml extension and it's working as intended, both codelens and duneDiagnostics.

File: ocaml-lsp-server/src/config_data.ml (hunk 1 above)
[the remark this answers is not in the document]
Comment (rgrinberg) [reply ocaml-lsp-server/src/config_data.ml:240]: The simpler format is fine, but we'd appreciate it if it was done in a separate PR.

File: ocaml-lsp-server/src/ocaml_lsp_server.ml [hunk 5]
```diff
@@ -592,17 +593,17 @@ let on_request :
...
     later (fun (_ : State.t) () -> Hover_req.handle rpc req mode) ()
   | TextDocumentReferences req -> later references req
   | TextDocumentCodeLensResolve codeLens -> now codeLens
-  | TextDocumentCodeLens req -> (
-    match state.configuration.data.codelens with
-    | Some { enable = true } -> later text_document_lens req
-    | _ -> now [])
+  | TextDocumentCodeLens req ->
+    if state.configuration.data.codelens.enable then
+      later text_document_lens req
```
Comment (rgrinberg) [inline ocaml-lsp-server/src/ocaml_lsp_server.ml:605]: These changes seem unrelated. Separate PR for them please.

File: ocaml-lsp-server/src/config_data.ml (hunk 1 above)
[the remark this answers is not in the document]
Comment (EduardoRFS) [reply ocaml-lsp-server/src/config_data.ml:240]: I placed the simpler type in a different commit, to make it easier to review. If you want I can remove it, but I would be glad if we can avoid bikeshedding here.

[the remark this answers is not in the document]
Comment (rgrinberg) [reply ocaml-lsp-server/src/config_data.ml:240]: If my comments are "bikeshedding", then you will have a difficult time contributing to this repository. I like to keep PR's focused on one of: bug fixing, new features, refactoring. So yes, keep it focused for the reviewers and move the unrelated stuff to separate PR's.

[the remark this answers is not in the document]
Comment (EduardoRFS) [reply ocaml-lsp-server/src/config_data.ml]: You could had dropped the commit yourself and merged. I removed the changes as you requested and I will not be making a separate PR.

Also I noticed that contributing will be annoying don't worry, like others I don't intended to contribute much here for this very specific reason.

**Comment by rgrinberg** [comment]
Alright, this is good to go after some minor modifications.

**Comment by EduardoRFS** [comment]
Thank you

**Comment by jakubrpawlowski** [comment]
Thank you so much for this update! 🙇‍♂️

**Comment by felipecrv** [comment]
I think this will improve my experience as well.
